plugins/corsair/fu-corsair-common.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2022 Andrii Dushko <andrii.dushko@developex.net>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "fu-corsair-common.h"
+
+FuCorsairDeviceKind
+fu_corsair_device_type_from_string(const gchar *kind)
+{
+	if (g_strcmp0(kind, "mouse") == 0)
+		return FU_CORSAIR_DEVICE_MOUSE;
+	if (g_strcmp0(kind, "receiver") == 0)
+		return FU_CORSAIR_DEVICE_DONGLE;
+	return FU_CORSAIR_DEVICE_UNKNOWN;
+}
+
+const gchar *
+fu_corsair_device_type_to_string(FuCorsairDeviceKind type)
+{
+	if (type == FU_CORSAIR_DEVICE_MOUSE)
+		return "mouse";
+	if (type == FU_CORSAIR_DEVICE_DONGLE)
+		return "receiver";
+
+	return "unknown";
+}
+
+guint32
+fu_corsair_calculate_crc(const guint8 *data, guint32 data_len)
+{
+	gboolean bit;
+	unsigned char c;
+	const unsigned char *d = (const unsigned char *)data;
+	guint32 crc = 0xffffffff;
+
+	while (data_len--) {
+		c = *d++;
+		for (guint i = 0x80; i > 0; i >>= 1) {
+			bit = crc & 0x80000000;
+			if (c & i) {
+				bit = !bit;
+			}
+			crc <<= 1;
+			if (bit) {
+				crc ^= 0x04c11db7;
+			}
+		}
+	}
+	return crc;
+}
+
+gchar *
+fu_corsair_parse_version(guint32 val)
```

Comment:
i think you can use `fu_common_version_from_uint32(val, FWUPD_VERSION_KIND_TRIPLET)` instead of this

Reply:
I can't use FWUPD_VERSION_KIND_TRIPLET because it expects bytes like this 0xAA.BB.CCDD, but versions on a device are stored in non-standard 0xCCDD.BB.AA order.
I'll leave a comment about it.

Reply:
Ahh, okay, I missed this nuance, sorry. Maybe rename your function fu_corsair_version_from_uint32 too?

Reply:
sure, I'll rename it